symengine/symengine_rcp.h

```diff
@@ -198,6 +198,56 @@ using Teuchos::print_stack_on_segfault;
 #endif
 
 
+class EnableRCPFromThis {
+// Public interface
+public:
+    unsigned int use_count() const {
+#if defined(WITH_SYMENGINE_RCP)
+        return refcount_;
+#else
+        return weak_self_ptr_.strong_count();
+#endif
+    }
+
+
+
+
+// Everything below is private interface
+public:
+#if defined(WITH_SYMENGINE_RCP)
+
+    //! Public variables if defined with SYMENGINE_RCP
+    // The reference counter is defined either as "unsigned int" (faster, but
+    // not thread safe) or as std::atomic<unsigned int> (slower, but thread
+    // safe). Semantically they are almost equivalent, except that the
+    // pre-decrement operator `operator--()` returns a copy for std::atomic
+    // instead of a reference to itself.
+    // The refcount_ is defined as mutable, because it does not change the
+    // state of the instance, but changes when more copies
+    // of the same instance are made.
+#  if defined(WITH_SYMENGINE_THREAD_SAFE)
+    mutable std::atomic<unsigned int> refcount_; // reference counter
+#  else
+    mutable unsigned int refcount_; // reference counter
+#  endif // WITH_SYMENGINE_THREAD_SAFE
+    EnableRCPFromThis() : refcount_(0) {}
+
+#else
+    mutable RCP<const EnableRCPFromThis> weak_self_ptr_;
+#endif // WITH_SYMENGINE_RCP
+
+    //! Get RCP<T> pointer to self (it will cast the pointer to T)
+    template <class T>
+    inline RCP<T> get_rcp_cast() const {
+#if defined(WITH_SYMENGINE_RCP)
+        return rcp(static_cast<T*>(this));
+#else
+        return rcp_static_cast<T>(weak_self_ptr_.create_strong());
```

Comment:
Good point. It's going to segfault. We need to implement debug time
checking for casts.

Sent from my mobile phone.
On Jul 15, 2015 12:29 PM, "Isuru Fernando" <notifications@github.com> wrote:

> In symengine/symengine_rcp.h
> <https://github.com/sympy/symengine/pull/532#discussion_r34712246>:
>
> > +#  else
> > +    mutable unsigned int refcount_; // reference counter
> > +#  endif // WITH_SYMENGINE_THREAD_SAFE
> > +    EnableRCPFromThis() : refcount_(0) {}
> > +
> > +#else
> > +    mutable RCP<const EnableRCPFromThis> weak_self_ptr_;
> > +#endif // WITH_SYMENGINE_RCP
> > +
> > +    //! Get RCP<T> pointer to self (it will cast the pointer to T)
> > +    template <class T>
> > +    inline RCP<T> get_rcp_cast() const {
> > +#if defined(WITH_SYMENGINE_RCP)
> > +        return rcp(static_cast<T*>(this));
> > +#else
> > +        return rcp_static_cast<T>(weak_self_ptr_.create_strong());
>
> This is from another PR, but what happens when you try to cast from Add
> to Mul in Debug mode?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/symengine/pull/532/files#r34712246>.
>